app/models/concerns/foreman/thread_session.rb

```diff
@@ -37,6 +37,24 @@ def clear_thread
       end
     end
 
+    # This allows getting and setting all current values in case it's needed,
```

Comment:
I can see that there are others that use `Thread.current` to store thread specific context.

May I suggest making the context a bit more generic? It should be pretty easy:
* instead of storing `:user`, `:organization` and `:location` directly in the thread, let's store a new hash there with these keys.
* Modify `*Model` classes to access `Thread.current[:foreman_context][:X]`
* Copying the context from one thread to another would be as simple as deep cloning the context hash.
* Either extend Thread class with `foreman_context` method that will access the new hash, or add this method to `Foreman::ThreadSession` module.

With these changes everyone will be sure that the data is passed to new threads.